eras/conway/impl/src/Cardano/Ledger/Conway/TxBody.hs

```diff
@@ -195,8 +202,18 @@ instance
       bodyFields 16 = ofield (\x tx -> tx {ctbrCollateralReturn = x}) From
       bodyFields 17 = ofield (\x tx -> tx {ctbrTotalCollateral = x}) From
       bodyFields 18 = field (\x tx -> tx {ctbrReferenceInputs = x}) From
-      bodyFields 19 = field (\x tx -> tx {ctbrVotingProcedures = x}) From
-      bodyFields 20 = field (\x tx -> tx {ctbrProposalProcedures = x}) From
+      bodyFields 19 =
+        fieldGuarded
+          "TxBody: Cannot decode empty VotingProcedures"
+          (Map.null . unVotingProcedures)
+          (\x tx -> tx {ctbrVotingProcedures = x})
+          From
+      bodyFields 20 =
+        fieldGuarded
+          "TxBody: Cannot decode empty ProposalProcedures"
```

Comment:
What do you say if we abstract the message into a separate function in order to reduce duplication and rephrase the error message a bit:
```suggestion
          emptyFailure "proposalProcedures" "non-empty"
```
```haskell
  where
    emptyFailure fieldName requirement =
      "TxBody: '" ++ fieldName ++ "' must be " ++ requirement ++ " when supplied"
```

This way `emptyFailure` can be used for all such fields including the `treasuryDonations`

Reply:
Great idea! Thanks! :raised_hand_with_fingers_splayed: